tests/src/MessageLayout.cpp

```diff
@@ -0,0 +1,123 @@
+#include "Application.hpp"
+#include "common/Literals.hpp"
+#include "controllers/accounts/AccountController.hpp"
+#include "controllers/commands/Command.hpp"
+#include "controllers/commands/CommandController.hpp"
+#include "controllers/hotkeys/HotkeyController.hpp"
+#include "messages/layouts/MessageLayout.hpp"
+#include "messages/MessageBuilder.hpp"
+#include "messages/MessageElement.hpp"
+#include "mocks/EmptyApplication.hpp"
+#include "singletons/Emotes.hpp"
+#include "singletons/Fonts.hpp"
+#include "singletons/Paths.hpp"
+#include "singletons/Settings.hpp"
+#include "singletons/Theme.hpp"
+#include "singletons/WindowManager.hpp"
+#include "widgets/Notebook.hpp"
+#include "widgets/splits/Split.hpp"
+#include "widgets/splits/SplitInput.hpp"
+
+#include <gmock/gmock.h>
+#include <gtest/gtest.h>
+#include <QDebug>
+#include <QString>
+
+#include <memory>
+
+using namespace chatterino;
+
+namespace {
+
+class MockApplication : mock::EmptyApplication
+{
+public:
+    MockApplication()
+        : settings(this->settingsDir.filePath("settings.json"))
+        , fonts(this->settings)
+        , windowManager(this->paths)
+    {
+    }
+    Theme *getThemes() override
+    {
+        return &this->theme;
+    }
+
+    HotkeyController *getHotkeys() override
+    {
+        return &this->hotkeys;
+    }
+
+    Fonts *getFonts() override
+    {
+        return &this->fonts;
+    }
+
+    WindowManager *getWindows() override
+    {
+        return &this->windowManager;
+    }
+
+    AccountController *getAccounts() override
+    {
+        return &this->accounts;
+    }
+
+    CommandController *getCommands() override
+    {
+        return &this->commands;
+    }
+
+    IEmotes *getEmotes() override
+    {
+        return &this->emotes;
+    }
+
+    Settings settings;
+    Theme theme;
+    HotkeyController hotkeys;
+    Fonts fonts;
+    Paths paths;
+    WindowManager windowManager;
+    AccountController accounts;
+    CommandController commands;
+    Emotes emotes;
+};
+
+constexpr int WIDTH = 300;
+
+class MessageLayoutTest
+{
+public:
+    // "aaaaaaaa bbbbbbbb cccccccc"
+    MessageLayoutTest(const QString &text)
+    {
+        MessageBuilder builder;
+        builder.append(
+            std::make_unique<TextElement>(text, MessageElementFlag::Text));
+        this->layout = std::make_unique<MessageLayout>(builder.release());
+        this->layout->layout(WIDTH, 1, MessageElementFlag::Text, false);
+    }
+
+    MockApplication mockApplication;
```

Comment:
I still don't quite understand how constructing `MockApplication` here makes it so any all to `getIApp` gets the MockApplication. Also this should probably only be done once instead of each time a test is created

Reply:
The constructor of MockApplication sets a singleton to point to itself (check mocks/EmptyApplication)